File: templates/ruflet_flutter_template/apple_packages/ruflet_apple/Sources/RufletEngine/Routing/route_information_provider.swift
import Foundation
import RufletProtocol

/// Route information carried between Apple's URL delivery APIs and Ruflet's
/// path-based router.
public struct RufletRouteInformation: Equatable, Sendable {
  public let uri: URL
  public let state: RufletValue?

  public init(uri: URL, state: RufletValue? = nil) {
    self.uri = uri
    self.state = state
  }
}

/// Normalizes external URLs such as `ruflet://ruflet-host/aaa` to `/aaa`,
/// matching FletRouteInformationProvider's path, query, and fragment contract.
@MainActor
public final class RufletRouteInformationProvider: ObservableObject {
  @Published public private(set) var value: RufletRouteInformation

  public init(initialRouteInformation: RufletRouteInformation) {
    value = Self.normalize(initialRouteInformation)
  }

  public static func normalize(
    _ routeInformation: RufletRouteInformation
  ) -> RufletRouteInformation {
    let source = URLComponents(
      url: routeInformation.uri,
      resolvingAgainstBaseURL: false)
    var normalized = URLComponents()
    normalized.percentEncodedPath = source?.percentEncodedPath.isEmpty == false
      ? source!.percentEncodedPath
      : "/"
    normalized.percentEncodedQuery = source?.percentEncodedQuery
    normalized.percentEncodedFragment = source?.percentEncodedFragment
    guard let uri = normalized.url else {
      preconditionFailure("Unable to normalize Ruflet route: \(routeInformation.uri)")
    }
    return RufletRouteInformation(uri: uri, state: routeInformation.state)
  }

  @discardableResult
  public func didPushRouteInformation(
    _ routeInformation: RufletRouteInformation
  ) -> Bool {
    setValue(routeInformation)
    return true
  }

  public func routerReportsNewRouteInformation(
    _ routeInformation: RufletRouteInformation
  ) {
    setValue(routeInformation)
  }

  private func setValue(_ routeInformation: RufletRouteInformation) {
    let normalized = Self.normalize(routeInformation)
    guard normalized != value else { return }
    value = normalized
  }
}

/// Local multi-view counterpart to the platform route provider.
public typealias RufletLocalRouteInformationProvider = RufletRouteInformationProvider

File: templates/ruflet_flutter_template/ruflet_packages/ruflet/lib/src/routing/route_information_provider.dart
import 'package:flutter/widgets.dart';

/// Normalizes external URIs (e.g. `ruflet://ruflet-host/aaa`) into the path-based
/// form used by Ruflet routing (e.g. `/aaa`).
class RufletRouteInformationProvider extends PlatformRouteInformationProvider {
  RufletRouteInformationProvider({
    required super.initialRouteInformation,
  });

  static RouteInformation normalize(RouteInformation routeInformation) {
    final uri = routeInformation.uri;
    return RouteInformation(
      uri: Uri(
        path: uri.path.isEmpty ? '/' : uri.path,
        query: uri.hasQuery ? uri.query : null,
        fragment: uri.hasFragment ? uri.fragment : null,
      ),
      state: routeInformation.state,
    );
  }

  @override
  Future<bool> didPushRouteInformation(RouteInformation routeInformation) {
    final normalized = normalize(routeInformation);
    debugPrint(
        "RufletRouteInformationProvider.didPushRouteInformation: ${routeInformation.uri} -> ${normalized.uri}");
    return super.didPushRouteInformation(normalized);
  }
}

class RufletLocalRouteInformationProvider extends RouteInformationProvider
    with ChangeNotifier {
  RouteInformation _value;

  RufletLocalRouteInformationProvider({
    required RouteInformation initialRouteInformation,
  }) : _value = RufletRouteInformationProvider.normalize(
            initialRouteInformation,
          );

  @override
  RouteInformation get value => _value;

  @override
  void routerReportsNewRouteInformation(
    RouteInformation routeInformation, {
    RouteInformationReportingType type = RouteInformationReportingType.none,
  }) {
    _setValue(routeInformation);
  }

  Future<bool> didPushRouteInformation(RouteInformation routeInformation) async {
    _setValue(routeInformation);
    return true;
  }

  void _setValue(RouteInformation routeInformation) {
    final normalized = RufletRouteInformationProvider.normalize(routeInformation);
    if (_value.uri == normalized.uri && _value.state == normalized.state) {
      return;
    }
    _value = normalized;
    notifyListeners();
  }
}
